src/fs/metadata_log_bootstrap.hh

```
@@ -0,0 +1,485 @@
+                case CHECKPOINT: // CHECKPOINT cannot appear as part of checkpointed data
+                    return invalid_entry_exception();
+                case NEXT_METADATA_CLUSTER:
+                    return bootstrap_next_metadata_cluster();
+                case CREATE_INODE:
+                    return bootstrap_create_inode();
+                case UPDATE_METADATA:
+                    return bootstrap_update_metadata();
+                case DELETE_INODE:
+                    return bootstrap_delete_inode();
+                case SMALL_WRITE:
+                    return bootstrap_small_write();
+                case MEDIUM_WRITE:
+                    return bootstrap_medium_write();
+                case LARGE_WRITE:
+                    return bootstrap_large_write();
+                case LARGE_WRITE_WITHOUT_MTIME:
+                    return bootstrap_large_write_without_mtime();
+                case TRUNCATE:
+                    return bootstrap_truncate();
+                case MTIME_UPDATE:
+                    return bootstrap_mtime_update();
+                case ADD_DIR_ENTRY:
+                    return bootstrap_add_dir_entry();
+                case CREATE_INODE_AS_DIR_ENTRY:
+                    return bootstrap_create_inode_as_dir_entry();
+                case RENAME_DIR_ENTRY:
+                    return bootstrap_rename_dir_entry();
+                case DELETE_DIR_ENTRY:
+                    return bootstrap_delete_dir_entry();
+                }
+
+                // unknown type => metadata log corruption
+                return invalid_entry_exception();
+            }).then([this] {
+                if (_curr_checkpoint.bytes_left() > 0) {
+                    return invalid_entry_exception(); // Corrupted checkpointed data
+                }
+                return now();
+            });
+        });
+    }
+
+    future<> bootstrap_next_metadata_cluster() {
+        ondisk_next_metadata_cluster entry;
```

We allow for multiple NEXT_METADATA_CLUSTER entries. I think that it is invalid so I would report an error in that case.